Support correlated scalar aggregation subqueries (v7)

kokosing opened the pull request:
Support correlated scalar aggregation subqueries
    
A scalar aggregation subquery aggregates all the rows into a
single group (empty GROUP BY clause). For example:
```
SELECT min(a) FROM t WHERE b > 0
```
    
This commit add support for correlated scalar aggregation subqueries
which have WHERE clause of conjunction of equality comparison.
For example:
```
SELECT (SELECT min(a) FROM t1(a,c) WHERE b = 0 AND c = d) FROM t2(b,d)
```
    
Logically such query is rewritten into left outer join where correlation
used in equality comparisons (from WHERE clause) are used as join
condition.

kokosing commented:
Supersedes https://github.com/prestodb/presto/pull/6155

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -230,6 +233,13 @@ private PlanBuilder appendExistSubqueryApplyNode(PlanBuilder subPlan, ExistsPred
 
+        Symbol exists = symbolAllocator.newSymbol("exists", BOOLEAN);
+        if (isAggregation(subqueryPlan)) {
+            // aggregation always return at least one row
```
This is not necessarily true. If it has a GROUP BY on anything other than (), it will return 0 rows if the input contains 0 rows:

```sql
SELECT count(*) 
FROM (VALUES 1,2,3,4) t(x) 
WHERE false 
GROUP BY x;
```

=>

```
 _col0
-------
(0 rows)
```

martint commented:
I merged everything but the "optimize EXISTS" commit and the one that adds documentation for subqueries.

kokosing commented:
@martint Rebased and addressed comments. Superseded by https://github.com/prestodb/presto/pull/6201.

Unfortunately, I didn't run all the tests before posting this PR yesterday. I just run subset of them. I wanted travis to run them all. As usual in such cases travis execution failed as there were some minor issues. Commits you merged yesterday are causing master build to fail. It is already solved by https://github.com/prestodb/presto/pull/6194. Sorry about that, I should explicitly write that you should wait for travis before merging. Anyway, thank you for merging!
